File: okay-swift/Sources/OkayProtocol/Wire.swift
// The JSON codec, by hand and by the document: a sum is an object with
// ONE key (the case name), a product its fields by name, an enumeration
// its lower-case name. Reading is TOTAL — damage is nil, never a
// throw — because a wire that dies on one bad line loses every good
// one after it. Foundation's JSONSerialization, nothing else.
import Foundation

public enum Wire {

    // ------------------------------------------------------------ write

    public static func line(_ m: Msg) -> String {
        let data = try! JSONSerialization.data(withJSONObject: msg(m), options: [.sortedKeys])
        return String(data: data, encoding: .utf8)!
    }

    public static func msg(_ m: Msg) -> [String: Any] {
        switch m {
        case let .hello(vocab, version): return sum("Hello", ["vocab": vocab, "version": version])
        case let .tree(u): return sum("Tree", ["ui": ui(u)])
        case let .patch(p): return sum("Patch", ["patch": patch(p)])
        case let .event(e): return sum("Event", ["event": event(e)])
        case .close: return sum("Close", [:])
        }
    }

    public static func ui(_ u: Ui) -> [String: Any] {
        switch u {
        case let .text(s, st): return sum("Text", ["s": s, "style": style(st)])
        case let .row(c, k): return sum("Row", ["children": c.map(ui), "key": k])
        case let .column(c, k): return sum("Column", ["children": c.map(ui), "key": k])
        case let .box(c, d, w, g, p, k):
            return sum("Box", ["children": c.map(ui), "dir": d.rawValue, "weights": w, "gap": g, "pad": p, "key": k])
        case let .image(src, alt): return sum("Image", ["src": src, "alt": alt])
        case let .button(l, k, r): return sum("Button", ["label": l, "key": k, "role": r.rawValue])
        case let .input(v, k, l, kind, live):
            return sum("Input", ["value": v, "key": k, "label": l, "kind": kind.rawValue, "live": live])
        case let .check(on, k, l): return sum("Check", ["on": on, "key": k, "label": l])
        case let .select(o, s, k): return sum("Select", ["options": o, "selected": s, "key": k])
        case let .scroll(c, k): return sum("Scroll", ["child": ui(c), "key": k])
        case let .form(f, s, k): return sum("Form", ["fields": f.map(ui), "submit": s, "key": k])
        case let .items(i, k): return sum("Items", ["items": i.map(ui), "key": k])
        case let .table(h, r, k): return sum("Table", ["header": h, "rows": r.map { $0.map(ui) }, "key": k])
        case let .tabs(l, s, p, k): return sum("Tabs", ["labels": l, "selected": s, "pages": p.map(ui), "key": k])
        case let .modal(t, b, k): return sum("Modal", ["title": t, "body": ui(b), "key": k])
        case let .disclosure(t, o, b, k): return sum("Disclosure", ["title": t, "open": o, "body": ui(b), "key": k])
        }
    }

    public static func event(_ e: Event) -> [String: Any] {
        switch e {
        case let .pressed(k): return sum("Pressed", ["key": k])
        case let .edited(k, v): return sum("Edited", ["key": k, "value": v])
        case let .toggled(k, on): return sum("Toggled", ["key": k, "on": on])
        case let .chosen(k, i): return sum("Chosen", ["key": k, "index": i])
        case let .key(ch): return sum("Key", ["ch": ch])
        case let .resized(w, h): return sum("Resized", ["w": w, "h": h])
        case .closed: return sum("Closed", [:])
        case let .submitted(k, edits): return sum("Submitted", ["key": k, "edits": edits.map(event)])
        }
    }

    public static func patch(_ p: Patch) -> [String: Any] {
        switch p {
        case let .replace(path, u): return sum("Replace", ["path": path, "ui": ui(u)])
        case let .setText(path, s): return sum("SetText", ["path": path, "s": s])
        case let .setValue(path, s): return sum("SetValue", ["path": path, "s": s])
        case let .setChecked(path, on): return sum("SetChecked", ["path": path, "on": on])
        case let .setSelected(path, i): return sum("SetSelected", ["path": path, "index": i])
        case let .remove(path, i): return sum("Remove", ["path": path, "index": i])
        case let .reorder(path, o): return sum("Reorder", ["path": path, "order": o])
        case let .insert(path, i, u): return sum("Insert", ["path": path, "index": i, "ui": ui(u)])
        }
    }

    static func style(_ s: Style) -> [String: Any] {
        ["bold": s.bold, "dim": s.dim, "tone": s.tone.rawValue, "size": s.size.rawValue]
    }
    static func sum(_ name: String, _ fields: [String: Any]) -> [String: Any] { [name: fields] }

    // ------------------------------------------------------------- read

    public static func parse(_ line: String) -> Msg? {
        guard let data = line.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) else { return nil }
        return readMsg(obj)
    }

    public static func readMsg(_ j: Any) -> Msg? {
        guard let (name, f) = sumOf(j) else { return nil }
        switch name {
        case "Hello":
            guard let v = f["vocab"] as? [String], let ver = int(f["version"]) else { return nil }
            return .hello(vocab: v, version: ver)
        case "Tree": guard let u = readUi(f["ui"]) else { return nil }; return .tree(ui: u)
        case "Patch": guard let p = readPatch(f["patch"]) else { return nil }; return .patch(patch: p)
        case "Event": guard let e = readEvent(f["event"]) else { return nil }; return .event(event: e)
        case "Close": return .close
        default: return nil
        }
    }

    public static func readUi(_ j: Any?) -> Ui? {
        guard let j = j, let (name, f) = sumOf(j) else { return nil }
        func kids(_ field: String) -> [Ui]? {
            guard let arr = f[field] as? [Any] else { return nil }
            var out: [Ui] = []
            for x in arr { guard let u = readUi(x) else { return nil }; out.append(u) }
            return out
        }
        func str(_ k: String) -> String { f[k] as? String ?? "" }
        switch name {
        case "Text":
            guard let s = f["s"] as? String else { return nil }
            return .text(s: s, style: readStyle(f["style"]))
        case "Row": guard let c = kids("children") else { return nil }; return .row(children: c, key: str("key"))
        case "Column": guard let c = kids("children") else { return nil }; return .column(children: c, key: str("key"))
        case "Box":
            guard let c = kids("children"), let d = Dir(rawValue: str("dir")) else { return nil }
            return .box(children: c, dir: d, weights: ints(f["weights"]) ?? [], gap: int(f["gap"]) ?? 0, pad: int(f["pad"]) ?? 0, key: str("key"))
        case "Image": guard let src = f["src"] as? String else { return nil }; return .image(src: src, alt: str("alt"))
        case "Button":
            guard let l = f["label"] as? String else { return nil }
            return .button(label: l, key: str("key"), role: Role(rawValue: str("role")) ?? .plain)
        case "Input":
            guard let v = f["value"] as? String else { return nil }
            return .input(value: v, key: str("key"), label: str("label"), kind: InputKind(rawValue: str("kind")) ?? .text, live: f["live"] as? Bool ?? false)
        case "Check": guard let on = f["on"] as? Bool else { return nil }; return .check(on: on, key: str("key"), label: str("label"))
        case "Select":
            guard let o = f["options"] as? [String] else { return nil }
            return .select(options: o, selected: int(f["selected"]) ?? 0, key: str("key"))
        case "Scroll": guard let c = readUi(f["child"]) else { return nil }; return .scroll(child: c, key: str("key"))
        case "Form":
            guard let fs = kids("fields"), let s = f["submit"] as? String else { return nil }
            return .form(fields: fs, submit: s, key: str("key"))
        case "Items": guard let i = kids("items") else { return nil }; return .items(items: i, key: str("key"))
        case "Table":
            guard let h = f["header"] as? [String], let rs = f["rows"] as? [Any] else { return nil }
            var rows: [[Ui]] = []
            for r in rs {
                guard let cells = r as? [Any] else { return nil }
                var row: [Ui] = []
                for c in cells { guard let u = readUi(c) else { return nil }; row.append(u) }
                rows.append(row)
            }
            return .table(header: h, rows: rows, key: str("key"))
        case "Tabs":
            guard let l = f["labels"] as? [String], let p = kids("pages") else { return nil }
            return .tabs(labels: l, selected: int(f["selected"]) ?? 0, pages: p, key: str("key"))
        case "Modal":
            guard let t = f["title"] as? String, let b = readUi(f["body"]) else { return nil }
            return .modal(title: t, body: b, key: str("key"))
        case "Disclosure":
            guard let t = f["title"] as? String, let b = readUi(f["body"]) else { return nil }
            return .disclosure(title: t, open: f["open"] as? Bool ?? false, body: b, key: str("key"))
        default: return nil
        }
    }

    public static func readEvent(_ j: Any?) -> Event? {
        guard let j = j, let (name, f) = sumOf(j) else { return nil }
        switch name {
        case "Pressed": guard let k = f["key"] as? String else { return nil }; return .pressed(key: k)
        case "Edited": guard let k = f["key"] as? String, let v = f["value"] as? String else { return nil }; return .edited(key: k, value: v)
        case "Toggled": guard let k = f["key"] as? String, let on = f["on"] as? Bool else { return nil }; return .toggled(key: k, on: on)
        case "Chosen": guard let k = f["key"] as? String, let i = int(f["index"]) else { return nil }; return .chosen(key: k, index: i)
        case "Key": guard let ch = f["ch"] as? String else { return nil }; return .key(ch: ch)
        case "Resized": guard let w = int(f["w"]), let h = int(f["h"]) else { return nil }; return .resized(w: w, h: h)
        case "Closed": return .closed
        case "Submitted":
            guard let k = f["key"] as? String, let es = f["edits"] as? [Any] else { return nil }
            var edits: [Event] = []
            for e in es { guard let ev = readEvent(e) else { return nil }; edits.append(ev) }
            return .submitted(key: k, edits: edits)
        default: return nil
        }
    }

    public static func readPatch(_ j: Any?) -> Patch? {
        guard let j = j, let (name, f) = sumOf(j), let path = ints(f["path"]) else { return nil }
        switch name {
        case "Replace": guard let u = readUi(f["ui"]) else { return nil }; return .replace(path: path, ui: u)
        case "SetText": guard let s = f["s"] as? String else { return nil }; return .setText(path: path, s: s)
        case "SetValue": guard let s = f["s"] as? String else { return nil }; return .setValue(path: path, s: s)
        case "SetChecked": guard let on = f["on"] as? Bool else { return nil }; return .setChecked(path: path, on: on)
        case "SetSelected": guard let i = int(f["index"]) else { return nil }; return .setSelected(path: path, index: i)
        case "Remove": guard let i = int(f["index"]) else { return nil }; return .remove(path: path, index: i)
        case "Reorder": guard let o = ints(f["order"]) else { return nil }; return .reorder(path: path, order: o)
        case "Insert": guard let i = int(f["index"]), let u = readUi(f["ui"]) else { return nil }; return .insert(path: path, index: i, ui: u)
        default: return nil
        }
    }

    static func readStyle(_ j: Any?) -> Style {
        guard let f = j as? [String: Any] else { return Style() }
        return Style(bold: f["bold"] as? Bool ?? false, dim: f["dim"] as? Bool ?? false,
                     tone: Tone(rawValue: f["tone"] as? String ?? "") ?? .plain,
                     size: Size(rawValue: f["size"] as? String ?? "") ?? .normal)
    }

    /// the one-key object a sum is
    static func sumOf(_ j: Any) -> (String, [String: Any])? {
        guard let o = j as? [String: Any], o.count == 1, let (k, v) = o.first, let f = v as? [String: Any] else { return nil }
        return (k, f)
    }
    static func int(_ j: Any?) -> Int? {
        if let n = j as? NSNumber { return n.intValue }
        return nil
    }
    static func ints(_ j: Any?) -> [Int]? {
        guard let arr = j as? [Any] else { return nil }
        var out: [Int] = []
        for x in arr { guard let n = int(x) else { return nil }; out.append(n) }
        return out
    }
}

File: okay-compose/protocol/src/main/kotlin/okay/compose/protocol/Wire.kt
// The JSON codec, by hand and by the document: a sum is an object
// with ONE key (the case name), a product its fields by name, an
// enumeration its lower-case name. Reading is TOTAL — damage is null,
// never a throw — because a wire that dies on one bad line loses
// every good one after it.
package okay.compose.protocol

import kotlinx.serialization.json.*

object Wire {
    private val json = Json { encodeDefaults = true }

    // ------------------------------------------------------------ write

    fun line(m: Msg): String = json.encodeToString(JsonElement.serializer(), msg(m))

    fun msg(m: Msg): JsonElement = when (m) {
        is Msg.Hello -> case("Hello", "vocab" to arr(m.vocab.map { JsonPrimitive(it) }), "version" to JsonPrimitive(m.version))
        is Msg.Tree -> case("Tree", "ui" to ui(m.ui))
        is Msg.Patch -> case("Patch", "patch" to patch(m.patch))
        is Msg.Event -> case("Event", "event" to event(m.event))
        Msg.Close -> case("Close")
    }

    fun ui(u: Ui): JsonElement = when (u) {
        is Ui.Text -> case("Text", "s" to str(u.s), "style" to style(u.style))
        is Ui.Row -> case("Row", "children" to uis(u.children), "key" to str(u.key))
        is Ui.Column -> case("Column", "children" to uis(u.children), "key" to str(u.key))
        is Ui.Box -> case(
            "Box", "children" to uis(u.children), "dir" to str(u.dir.wire),
            "weights" to arr(u.weights.map { JsonPrimitive(it) }), "gap" to JsonPrimitive(u.gap),
            "pad" to JsonPrimitive(u.pad), "key" to str(u.key),
        )
        is Ui.Image -> case("Image", "src" to str(u.src), "alt" to str(u.alt))
        is Ui.Button -> case("Button", "label" to str(u.label), "key" to str(u.key), "role" to str(u.role.name.lowercase()))
        is Ui.Input -> case(
            "Input", "value" to str(u.value), "key" to str(u.key), "label" to str(u.label),
            "kind" to str(u.kind.name.lowercase()), "live" to JsonPrimitive(u.live),
        )
        is Ui.Check -> case("Check", "on" to JsonPrimitive(u.on), "key" to str(u.key), "label" to str(u.label))
        is Ui.Select -> case("Select", "options" to arr(u.options.map { str(it) }), "selected" to JsonPrimitive(u.selected), "key" to str(u.key))
        is Ui.Scroll -> case("Scroll", "child" to ui(u.child), "key" to str(u.key))
        is Ui.Form -> case("Form", "fields" to uis(u.fields), "submit" to str(u.submit), "key" to str(u.key))
        is Ui.Items -> case("Items", "items" to uis(u.items), "key" to str(u.key))
        is Ui.Table -> case("Table", "header" to arr(u.header.map { str(it) }), "rows" to arr(u.rows.map { uis(it) }), "key" to str(u.key))
        is Ui.Tabs -> case("Tabs", "labels" to arr(u.labels.map { str(it) }), "selected" to JsonPrimitive(u.selected), "pages" to uis(u.pages), "key" to str(u.key))
        is Ui.Modal -> case("Modal", "title" to str(u.title), "body" to ui(u.body), "key" to str(u.key))
        is Ui.Disclosure -> case("Disclosure", "title" to str(u.title), "open" to JsonPrimitive(u.open), "body" to ui(u.body), "key" to str(u.key))
    }

    fun event(e: Event): JsonElement = when (e) {
        is Event.Pressed -> case("Pressed", "key" to str(e.key))
        is Event.Edited -> case("Edited", "key" to str(e.key), "value" to str(e.value))
        is Event.Toggled -> case("Toggled", "key" to str(e.key), "on" to JsonPrimitive(e.on))
        is Event.Chosen -> case("Chosen", "key" to str(e.key), "index" to JsonPrimitive(e.index))
        is Event.Key -> case("Key", "ch" to str(e.ch))
        is Event.Resized -> case("Resized", "w" to JsonPrimitive(e.w), "h" to JsonPrimitive(e.h))
        Event.Closed -> case("Closed")
        is Event.Submitted -> case("Submitted", "key" to str(e.key), "edits" to arr(e.edits.map { event(it) }))
    }

    fun patch(p: Patch): JsonElement = when (p) {
        is Patch.Replace -> case("Replace", "path" to ints(p.path), "ui" to ui(p.ui))
        is Patch.SetText -> case("SetText", "path" to ints(p.path), "s" to str(p.s))
        is Patch.SetValue -> case("SetValue", "path" to ints(p.path), "s" to str(p.s))
        is Patch.SetChecked -> case("SetChecked", "path" to ints(p.path), "on" to JsonPrimitive(p.on))
        is Patch.SetSelected -> case("SetSelected", "path" to ints(p.path), "index" to JsonPrimitive(p.index))
        is Patch.Remove -> case("Remove", "path" to ints(p.path), "index" to JsonPrimitive(p.index))
        is Patch.Reorder -> case("Reorder", "path" to ints(p.path), "order" to ints(p.order))
        is Patch.Insert -> case("Insert", "path" to ints(p.path), "index" to JsonPrimitive(p.index), "ui" to ui(p.ui))
    }

    private fun style(s: Style): JsonElement = JsonObject(
        mapOf(
            "bold" to JsonPrimitive(s.bold), "dim" to JsonPrimitive(s.dim),
            "tone" to str(s.tone.name.lowercase()), "size" to str(s.size.name.lowercase()),
        ),
    )

    private fun case(name: String, vararg fields: Pair<String, JsonElement>): JsonElement =
        JsonObject(mapOf(name to JsonObject(fields.toMap())))
    private fun str(s: String) = JsonPrimitive(s)
    private fun arr(xs: List<JsonElement>) = JsonArray(xs)
    private fun ints(xs: List<Int>) = JsonArray(xs.map { JsonPrimitive(it) })
    private fun uis(xs: List<Ui>) = JsonArray(xs.map { ui(it) })

    // ------------------------------------------------------------- read

    fun parse(line: String): Msg? = try {
        readMsg(json.parseToJsonElement(line))
    } catch (_: Exception) {
        null
    }

    fun readMsg(j: JsonElement): Msg? {
        val (name, f) = sum(j) ?: return null
        return when (name) {
            "Hello" -> Msg.Hello(f.strings("vocab") ?: return null, f.int("version") ?: return null)
            "Tree" -> Msg.Tree(readUi(f["ui"] ?: return null) ?: return null)
            "Patch" -> Msg.Patch(readPatch(f["patch"] ?: return null) ?: return null)
            "Event" -> Msg.Event(readEvent(f["event"] ?: return null) ?: return null)
            "Close" -> Msg.Close
            else -> null
        }
    }

    fun readUi(j: JsonElement): Ui? {
        val (name, f) = sum(j) ?: return null
        fun kids(field: String): List<Ui>? = (f[field] as? JsonArray)?.map { readUi(it) ?: return null }
        return when (name) {
            "Text" -> Ui.Text(f.str("s") ?: return null, readStyle(f["style"]) ?: return null)
            "Row" -> Ui.Row(kids("children") ?: return null, f.str("key") ?: "")
            "Column" -> Ui.Column(kids("children") ?: return null, f.str("key") ?: "")
            "Box" -> Ui.Box(
                kids("children") ?: return null,
                when (f.str("dir")) { "h" -> Dir.Horizontal; "v" -> Dir.Vertical; else -> return null },
                f.ints("weights") ?: emptyList(), f.int("gap") ?: 0, f.int("pad") ?: 0, f.str("key") ?: "",
            )
            "Image" -> Ui.Image(f.str("src") ?: return null, f.str("alt") ?: "")
            "Button" -> Ui.Button(f.str("label") ?: return null, f.str("key") ?: "", enum<Role>(f.str("role")) ?: Role.Plain)
            "Input" -> Ui.Input(
                f.str("value") ?: return null, f.str("key") ?: "", f.str("label") ?: "",
                enum<InputKind>(f.str("kind")) ?: InputKind.Text, f.bool("live") ?: false,
            )
            "Check" -> Ui.Check(f.bool("on") ?: return null, f.str("key") ?: "", f.str("label") ?: "")
            "Select" -> Ui.Select(f.strings("options") ?: return null, f.int("selected") ?: 0, f.str("key") ?: "")
            "Scroll" -> Ui.Scroll(readUi(f["child"] ?: return null) ?: return null, f.str("key") ?: "")
            "Form" -> Ui.Form(kids("fields") ?: return null, f.str("submit") ?: return null, f.str("key") ?: "")
            "Items" -> Ui.Items(kids("items") ?: return null, f.str("key") ?: "")
            "Table" -> Ui.Table(
                f.strings("header") ?: return null,
                (f["rows"] as? JsonArray)?.map { row -> (row as? JsonArray)?.map { readUi(it) ?: return null } ?: return null } ?: return null,
                f.str("key") ?: "",
            )
            "Tabs" -> Ui.Tabs(f.strings("labels") ?: return null, f.int("selected") ?: 0, kids("pages") ?: return null, f.str("key") ?: "")
            "Modal" -> Ui.Modal(f.str("title") ?: return null, readUi(f["body"] ?: return null) ?: return null, f.str("key") ?: "")
            "Disclosure" -> Ui.Disclosure(
                f.str("title") ?: return null, f.bool("open") ?: false,
                readUi(f["body"] ?: return null) ?: return null, f.str("key") ?: "",
            )
            else -> null
        }
    }

    fun readEvent(j: JsonElement): Event? {
        val (name, f) = sum(j) ?: return null
        return when (name) {
            "Pressed" -> Event.Pressed(f.str("key") ?: return null)
            "Edited" -> Event.Edited(f.str("key") ?: return null, f.str("value") ?: return null)
            "Toggled" -> Event.Toggled(f.str("key") ?: return null, f.bool("on") ?: return null)
            "Chosen" -> Event.Chosen(f.str("key") ?: return null, f.int("index") ?: return null)
            "Key" -> Event.Key(f.str("ch") ?: return null)
            "Resized" -> Event.Resized(f.int("w") ?: return null, f.int("h") ?: return null)
            "Closed" -> Event.Closed
            "Submitted" -> Event.Submitted(
                f.str("key") ?: return null,
                (f["edits"] as? JsonArray)?.map { readEvent(it) ?: return null } ?: return null,
            )
            else -> null
        }
    }

    fun readPatch(j: JsonElement): Patch? {
        val (name, f) = sum(j) ?: return null
        val path = f.ints("path") ?: return null
        return when (name) {
            "Replace" -> Patch.Replace(path, readUi(f["ui"] ?: return null) ?: return null)
            "SetText" -> Patch.SetText(path, f.str("s") ?: return null)
            "SetValue" -> Patch.SetValue(path, f.str("s") ?: return null)
            "SetChecked" -> Patch.SetChecked(path, f.bool("on") ?: return null)
            "SetSelected" -> Patch.SetSelected(path, f.int("index") ?: return null)
            "Remove" -> Patch.Remove(path, f.int("index") ?: return null)
            "Reorder" -> Patch.Reorder(path, f.ints("order") ?: return null)
            "Insert" -> Patch.Insert(path, f.int("index") ?: return null, readUi(f["ui"] ?: return null) ?: return null)
            else -> null
        }
    }

    private fun readStyle(j: JsonElement?): Style? {
        val f = j as? JsonObject ?: return Style()
        return Style(
            f.bool("bold") ?: false, f.bool("dim") ?: false,
            enum<Tone>(f.str("tone")) ?: Tone.Plain, enum<Size>(f.str("size")) ?: Size.Normal,
        )
    }

    /** the one-key object a sum is */
    private fun sum(j: JsonElement): Pair<String, JsonObject>? {
        val o = j as? JsonObject ?: return null
        if (o.size != 1) return null
        val (k, v) = o.entries.first()
        return k to (v as? JsonObject ?: return null)
    }

    private inline fun <reified E : Enum<E>> enum(name: String?): E? =
        name?.let { n -> enumValues<E>().firstOrNull { it.name.lowercase() == n } }
    private fun JsonObject.str(k: String): String? = (this[k] as? JsonPrimitive)?.takeIf { it.isString }?.content
    private fun JsonObject.bool(k: String): Boolean? = (this[k] as? JsonPrimitive)?.booleanOrNull
    private fun JsonObject.int(k: String): Int? = (this[k] as? JsonPrimitive)?.doubleOrNull?.toInt()
    private fun JsonObject.ints(k: String): List<Int>? = (this[k] as? JsonArray)?.map { (it as? JsonPrimitive)?.doubleOrNull?.toInt() ?: return null }
    private fun JsonObject.strings(k: String): List<String>? =
        (this[k] as? JsonArray)?.map { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content ?: return null }
}
